### PR TITLE
support network-3.0 (fixes #180)

### LINKED ISSUE #180: lambdabot is incompatible with network-2.7
See

  https://travis-ci.org/lambdabot/lambdabot/jobs/524511492

Probable cause: network-3.0 is incompatible with earlier versions. Will investigate tomorrow.

### PR DESCRIPTION
**Pull request by int-e**

### CODE REVIEW

File: lambdabot-haskell-plugins/lambdabot-haskell-plugins.cabal [hunk 1]
```diff
@@ -98,7 +98,7 @@ library
                         containers              >= 0.4,
                         directory               >= 1.1,
                         filepath                >= 1.3,
-                        haskell-src-exts-simple >= 1.21.0.0 && < 1.22,
+                        haskell-src-exts-simple >= 1.18.0.0 && < 1.22,
```
Comment (int-e) [inline lambdabot-haskell-plugins/lambdabot-haskell-plugins.cabal]: The point is that a) there was no reason to bump the lower bound and b) there was no viable install plan for ghc-7.10 anymore.

**Comment by int-e** [comment]
All green now. Comments welcome!

File: lambdabot-core/lambdabot-core.cabal [hunk 2]
```diff
@@ -110,3 +110,7 @@ library
 
   if !os(windows)
     build-depends:      unix                    >= 2.5
+
+  -- work around a problem with building hoogle on ghc-7.10.3
+  if impl(ghc < 8)
+    build-depends:      network                 < 3.0
```
Comment (int-e) [inline lambdabot-core/lambdabot-core.cabal]: See also https://github.com/ndmitchell/hoogle/issues/296 which should make this superfluous.

File: lambdabot-haskell-plugins/lambdabot-haskell-plugins.cabal [hunk 3]
```diff
@@ -117,7 +117,7 @@ library
                         -- runtime dependencies - for eval etc.
                         arrows                  >= 0.4,
                         data-memocombinators    >= 0.4,
-                        hoogle                  >= 4.2,
+                        hoogle                  >= 5.0.17.1,
```
Comment (int-e) [inline lambdabot-haskell-plugins/lambdabot-haskell-plugins.cabal]: Ironically, this implies a network >= 2.6 constraint. But that's no big deal, I think.

Comment (int-e) [reply lambdabot-haskell-plugins/lambdabot-haskell-plugins.cabal]: In fact network-2.6 predates ghc-7.10.1.

**Comment by int-e** [comment]
@peti any comments before I push this?

**Comment by peti** [comment]
The changes look fine to me insofar as they get the job done. I would argue, though, that supporting old versions of `network` through glue code is not something I'd invest time into, though. I'd rather port the code to network 3.x, put a lower version bound on that library, and never look back. That's what I have done in my other projects, at least. Just my 2 cents ...

**Comment by int-e** [comment]
There really is no glue code. The new code in `Lambdabot.Util.Network` provides the `connectTo` functionality that was completely dropped from the network package. The only concession to older versions of the network package is the prime in the name `connectTo'`.

**Comment by int-e** [comment]
Let me rephrase: There is glue code, but its primary purpose is to ease porting the rest of lambdabot to the newer network package, not compatibility with older versions.

I've tried to find packages providing `connectTo` functionality on hackage, but I've only found
https://hackage.haskell.org/package/network-simple which currently doesn't support network>=3.

PS: Thanks for the comments!

**Comment by peti** [comment]
@int-e, can you actually compile `lambdabot-core` with `network < 3`? It seems to me like the unconditional dependency on `network-bsd` forces use of `network > 3`.

**Comment by int-e** [comment]
@peti: thanks!

Yes, network < 3 works by selecting different version of network-bsd:

  * network-bsd-2.8.1.0  requires network >= 3 && < 3.0.2
  * network-bsd-2.8.0.0  requires network >= 2.8 && < 2.8.1
  * network-bsd-2.7.0.0  requires network >= 2.7 && < 2.7.1

I've accidently excluded network-bsd-2.7, I will fix that.
